Support compilation on Windows

Timmmm opened the pull request:
Support compilation on Windows by disabling the REPL because the `linenoise` library doesn't compile on Windows.

This also requires manually setting `LEMLIB` due to the way OPAM works (or doesn't) on Windows.

Alasdair commented:
Could try setting `SAIL_PLUGIN_DIR` if `dune_site` is causing issues on Windows.

Alasdair commented:
For some reason the conditional compilation does not seem to be working quite right. The interpreter is also disabled on Linux, which is why the tests are failing.

Alasdair commented on src/bin/dune:
```diff
@@ -15,7 +15,14 @@
+  (select
+   repl.ml
+   from
+   (linenoise -> repl.enabled.ml)
```
What happens to the tests if we change this from `(linenoise -> repl.enabled.ml)` to `(!linenoise -> repl.disabled.ml)` and flip the fallback case too?

Alasdair replied:
Assuming they continue to fail after this CI run... bit of a shot in the dark

Timmmm replied:
Ah it was actually because the selection syntax was totally wrong - I was using `(= %os_type "Unix")` but apparently `%os_type` only works with `enable_if` and you need to use `(<> :os "win32")`. Seems to work now.

Timmmm commented:
Ok with some more commits (not yet in this PR) I managed to make this work. [Here's a release build](https://github.com/Timmmm/sail/actions/runs/18629960648/artifacts/4310857974) (it's a zip within a zip because Github zips all artifacts).

Needs some more work on the `make tarball` flow though - apparently there's no sane way to write a cross-platform Makefile if you want to do stuff like `rm -rf` and `cp`, so I did... a thing to work around that which I'm not very happy with. I think I'll just move the entire release preparation flow into OCaml.

Also one minor issue that I think probably affects all platforms is that `sail_manifest` gets packaged, which doesn't really make sense. I couldn't figure out how to stop that but it doesn't matter too much I guess. (Also I renamed it to `sail_maker` because of other stuff I added to it.)

Mercifully Smart Screen apparently doesn't seem apply to programs run from the command line so we don't have to worry about signing or buying certificates. @edolnx FYI

Alasdair commented:
I think at some point the `sail_manifest` tool was called `sail_build_tool`, as that was kind of the original idea. Because of how opam builds are sandboxed, building a small OCaml program was much more reliable than relying on shell scripts.

Timmmm commented:
Ok this seems to work now. Quite a lot of changes to the Github CI flows, sorry! I can probably break it down into a couple of smaller PRs if necessary (but I'd rather be lazy and not!).

Timmmm committed:
Support Windows compilation by disabling the REPL on Windows

This also currently includes some incidental CI improvements that I made when trying to punch Github CI in the face until it worked:

* Use explicit `os` versions
* Use `runner.os ==` instead of `startswith(matrix.os`
* Use the `setup-z3` action to set up Z3
* Use the `setup-ocaml` action. This includes caching.

Also some of the tarball generation logic is moved from the Makefile into `sail_maker` (OCaml) because it's difficult to make portable Makefiles.

Timmmm commented:
Sorry I didn't see that this hadn't passed CI. Looks happier now!

Alasdair commented:
Looks good! Should probably get a windows machine so I can really test myself.

I tried dune-ifying the lem build a while ago and I got like 90% of the way there but got stuck on the library setup. I think that's the main thing needed for it to work nicely.

Timmmm commented:
Thanks! I also just bought a Mac, so I'll have a go at doing that too at some point.
